### PR TITLE
further model testing

### PR DESCRIPTION
**Pull request by promatty**
trained models on KNN, Decision Trees, Random Forests, and Gradient Boosting, all with and without Cross Validation.

### CODE REVIEW

File: analysis/furtherModelTesting.ipynb [hunk 1]
```diff
@@ -0,0 +1,687 @@
...
+       "\n",
+       "    .dataframe thead th {\n",
+       "        text-align: right;\n",
+       "    }\n",
+       "</style>\n",
+       "<table border=\"1\" class=\"dataframe\">\n",
+       "  <thead>\n",
+       "    <tr style=\"text-align: right;\">\n",
+       "      <th></th>\n",
+       "      <th>secondsdifference</th>\n",
```
Comment (justin-phxm) [inline analysis/furtherModelTesting.ipynb]: I would be careful about this secondsdifference feature...
This is an engineered feature by comparing the timestamp of the previous packet with the timestamp of the current packet. Since we are cleaning the data (removing rows with null values), it does not make sense to keep this feature because the previous row from which this feature was computed from may not even exist. 

In addition, the secondsdifference of a packet should have NO EFFECT on the distance of Helios.

File: analysis/furtherModelTesting.ipynb [hunk 2]
```diff
@@ -0,0 +1,687 @@
...
+     "output_type": "stream",
+     "text": [
+      "_knn for KNN is: 0.9679120129289978\n",
+      "MAE for KNN is: 0.5085657314801215\n"
+     ]
+    }
+   ],
+   "source": [
+    "#split into training and test sets\n",
+    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=69420)\n",
```
Comment (justin-phxm) [inline analysis/furtherModelTesting.ipynb]: Declare a global variable for random_state=69420 instead of defining it in every single function. Same with test_size

Comment (justin-phxm) [reply analysis/furtherModelTesting.ipynb]: Not sure if you've seen or thought about this much, but train_test_split has a `shuffle` argument that is by default `true`. What do you think this does and why do you think it's necessary?

**Review by justin-phxm — CHANGES_REQUESTED** [review]

File: analysis/furtherModelTesting.ipynb (hunk 1 above)
Comment (justin-phxm) [reply analysis/furtherModelTesting.ipynb]: Do you think we should remove it from the data?

**Comment by justin-phxm** [comment]
What are the differences between the two metrics, MAE and RMSE? How do they relate to precision and accuracy? How can we explain this in layman's terms, apply it directly to Helios, and make real world decisions from this?

File: analysis/furtherModelTesting.ipynb (hunk 1 above)
Comment (promatty) [reply analysis/furtherModelTesting.ipynb]: Yeah it does make sense to just drop it then. I'll do that

File: analysis/furtherModelTesting.ipynb [hunk 3]
```diff
@@ -0,0 +1,1039 @@
...
+    "import pandas as pd\n",
+    "import numpy as np\n",
+    "from sklearn.neighbors import KNeighborsRegressor\n",
+    "from sklearn.tree import DecisionTreeRegressor\n",
+    "from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor\n",
+    "from sklearn.metrics import mean_squared_error, mean_absolute_error, make_scorer\n",
+    "from sklearn.model_selection import GridSearchCV, KFold, train_test_split\n",
+    "\n",
+    "#global variables for randomness seed and test size\n",
+    "global_random_state = 69420\n",
```
Comment (justin-phxm) [inline analysis/furtherModelTesting.ipynb]: Since these are constants, can we do this in ALL CAPS... GLOBAL_RANDOM_STATE=...

File: analysis/furtherModelTesting.ipynb [hunk 4]
```diff
@@ -0,0 +1,1039 @@
...
+    "#HYPERPARAMTER TUNING FOR KNN\n",
+    "from sklearn.model_selection import GridSearchCV\n",
+    "\n",
+    "#split into training and test sets\n",
+    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=global_test_size, random_state=global_random_state)\n",
+    "\n",
+    "#train the model with KNN\n",
+    "#knn_model = KNeighborsRegressor(n_neighbors=5)\n",
+    "# Define the parameter grid\n",
+    "param_grid = {'n_neighbors': range(1, 21)}\n",
```
Comment (justin-phxm) [inline analysis/furtherModelTesting.ipynb]: range does not need to be so large. This slows down the computation and from the graph it's easy to see that anything past like 15 n_neighbours has a worse effect.

However, I believe this is proportional to the size of our data. Once Helios is finished, we will need to update ALL our parameters to match the new data we will be receiving.

File: analysis/furtherModelTesting.ipynb [hunk 5]
```diff
@@ -0,0 +1,1039 @@
...
+      "MAE for Random Forests is: 0.9348572302396694\n"
+     ]
+    }
+   ],
+   "source": [
+    "# HYPERPARAMETER TUNING FOR RANDOM FORESTS\n",
+    "\n",
+    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=global_test_size, random_state=global_random_state)\n",
+    "# Define the parameter grid for Random Forest\n",
+    "param_grid_rf = {\n",
```
Comment (justin-phxm) [inline analysis/furtherModelTesting.ipynb]: Tell me how many times this model will be trained with this size of param_grid

File: analysis/furtherModelTesting.ipynb [hunk 6]
```diff
@@ -0,0 +1,1039 @@
...
+     "text": [
+      "RMSE for Random Forests is: 1.7014263662227194\n",
+      "MAE for Random Forests is: 0.9749912091165533\n"
+     ]
+    }
+   ],
+   "source": [
+    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=global_test_size, random_state=global_random_state)\n",
+    "\n",
+    "rf_model = RandomForestRegressor(random_state=global_random_state)\n",
```
Comment (justin-phxm) [inline analysis/furtherModelTesting.ipynb]: Why are we doing this again? You did it above with the grid_search

File: analysis/furtherModelTesting.ipynb (hunk 5 above)
Comment (justin-phxm) [reply analysis/furtherModelTesting.ipynb]: How would this change if we also wanted to use Cross validation of say... 5?

**Review by justin-phxm — CHANGES_REQUESTED** [review]
Few small changes.

Great conclusion.

**Comment by justin-phxm** [comment]
Also I was hoping you would do the hyper parameter tuning optimizations in `ML-12-optimize-model-performance-through-hyperparameter-tuning` but this is fine ig.

I feel more comfortable reviewing smaller PRs

File: analysis/furtherModelTesting.ipynb (hunk 5 above)
Comment (promatty) [reply analysis/furtherModelTesting.ipynb]: the model is trained once for each combination, so it's 3x4x3x3=108 times. cross validation of 5 means we would train 5 times per combination so 5x108= 540. Is this too much?

**Review by justin-phxm — APPROVED** [review]
